CloudWatch Agent integration for nvidia-training

annapoornanarayan opened the pull request:
*Issue #, if available:*

*Description of changes:*
NOTE: This will only work after CW Infra is also successfully deployed (https://github.com/aws/aws-k8s-tester/pull/648)
Changes include: 
- Added amiVariant, teamIdentifier, and KubernetesVersion flags with default versions enabled
- CloudWatch agent manifest configured to scrape Prometheus compatible DCGM GPU metrics with proper labeling
- Small change in DCGM to ensure headless service discovery
- Created additional function to scrape Region from node labels
- Applied manifest in main_test.go after rendering with required parameters once DCGM Exporter runs

Testing:

- Verified CloudWatch agent deploys successfully with pods showing that 90+ metrics are collected successfully
- Metrics are actively visible under ContainerInsights/Prometheus under the dimensions ["TestType", "KubernetesVersion", "Variant", "InstanceType", "TeamIdentifier"]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

annapoornanarayan commented:
Changed scrape interval and improved CW Config for clarity

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -156,6 +191,20 @@ func checkNodeTypes(ctx context.Context, config *envconf.Config) (context.Contex
+	if err != nil {
+		return "", fmt.Errorf("failed to list nodes: %w", err)
+	}
+	az := nodes.Items[0].Labels["topology.kubernetes.io/zone"]
```
Should use `topology.kubernetes.io/region` label directly instead of doing some parse based on zone label

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -31,26 +32,60 @@ func TestMain(m *testing.M) {
+	region, err := getRegionFromNodes(ctx, cfg)
+	if err != nil {
+		log.Printf("Warning: failed to get region from nodes, using default us-west-2: %v", err)
+		region = "us-west-2"
```
What will happen to CW agent If the test is not running in PDX but failed to get region info here and use PDX by defaut?

wwvela replied:
Yeah. I think should give some context about it in log like where the metric will be emitted to. like `Warning: failed to get region from nodes, the test metrics will be emitted to PDX by default`

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -31,26 +32,60 @@ func TestMain(m *testing.M) {
 
+	// Validate and transform flags
+	*amiVariant = strings.ToLower(*amiVariant)
+	if *kubernetesVersion != "1.32" && *kubernetesVersion != "1.33" {
+		log.Fatalf("kubernetesVersion must be either 1.32 or 1.33, got: %s", *kubernetesVersion)
+	}
```
version limit is not needed. We are able to run the test with other versions if we wanted and shouldn't avoid it. We only run 1.32 and 1.33 in our pipeline just because of limited GPU instance capacity resources.

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -31,26 +32,60 @@ func TestMain(m *testing.M) {
+		renderedCloudWatchAgentManifest,
 	}
 
+	log.Printf("Using template variables: REGION=%s, VERSION=%s, VARIANT=%s, INSTANCE_TYPE=%s, TEAM_IDENTIFIER=%s, TEST_NAME=nvidia.training", 
+		region, manifestVersion, *amiVariant, *nodeType, *teamIdentifier)
+
```
nit: didn't see very helpful to print these out in log, we can saw these info in metric dimension directly

wwvela commented on test/cases/nvidia-training/vars.go:
```diff
@@ -24,4 +27,7 @@ func init() {
 	nodeType = flag.String("nodeType", "", "Instance type for cluster nodes")
+	kubernetesVersion = flag.String("kubernetesVersion", "1.32", "Kubernetes version for the cluster")
+	amiVariant = flag.String("amiVariant", "al2023", "AMI variant for the cluster nodes")
+	teamIdentifier = flag.String("teamIdentifier", "node-runtime", "Team identifier for resource tagging")
```
The teamIdentifier default value should set to `unknown`, as someone who run this test in local and forget to pass this flag will emit their test metric datapoint in our team or any other team metric dashboards.

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -31,26 +32,60 @@ func TestMain(m *testing.M) {
+	// template data for CloudWatch Agent manifest
+	templateData := map[string]string{
+		"VERSION":       manifestVersion,
+		"VARIANT":       *amiVariant,
```
can pass `strings.ToLower(*amiVariant)` directly here

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -31,26 +32,60 @@ func TestMain(m *testing.M) {
+
+	// template data for CloudWatch Agent manifest
+	templateData := map[string]string{
+		"VERSION":       manifestVersion,
```
can pass `strings.Replace(*kubernetesVersion, ".", "", 1)` directly

annapoornanarayan commented:
Changes: 
- Used topology.kubernetes.io/region label directly
- Failure on region info if not found
- Better k8 version validation
- Clean up logs
- "unknown" default value for teamIdentifier
- Passed in validation of flags directly to map

annapoornanarayan commented on test/cases/nvidia-training/vars.go:
```diff
@@ -14,6 +14,9 @@ var (
 	bertTrainingImage *string
 	efaEnabled        *bool
 	nodeType          *string
+	kubernetesVersion *string
+	amiVariant	  *string
+	teamIdentifier    *string
```
Spoke offline - making separate PR for this change

mselim00 replied:
Discussed more offline, but to summarize, we should try to make this more generic. Rather than specific flags like `teamIdentifer`, we should consider a more generic one-off flag like `--metric-dimensions` that allows end users to set the desired dimensions they want for their use case. Letting the tests append one reserved dimension like `TEST_NAME` seems fine though since that's generally reusable. We also need to be careful with letting the tester assume certain set up was done in the deployer because there's no guarantee of a specific deployer being used to run the tester, so assumptions should be limited and carefully handled

mselim00 commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -156,6 +185,20 @@ func checkNodeTypes(ctx context.Context, config *envconf.Config) (context.Contex
 }
+
+// getRegionFromNodes extracts the AWS region from node labels
+func getRegionFromNodes(ctx context.Context, config *envconf.Config) (string, error) {
+	clientset, err := kubernetes.NewForConfig(config.Client().RESTConfig())
+	if err != nil {
+		return "", fmt.Errorf("failed to create Kubernetes client: %w", err)
+	}
+	nodes, err := clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{Limit: 1})
+	if err != nil {
+		return "", fmt.Errorf("failed to list nodes: %w", err)
+	}
+	return nodes.Items[0].Labels["topology.kubernetes.io/region"], nil
+}
```
this is fine, but an alternative would have been using the AWS config
```
import "github.com/aws/aws-sdk-go-v2/config"
...
if cfg, err := config.LoadDefaultConfig(context.TODO()); err != nil {
   return "", fmt.Errorf("failed loading config, %v", err)
else {
   return cfg.Region, nil
}
```
though since we don't currently use any AWS APIs in this test, there's been no expectation that the environment is configured correctly. I think it could be worth adding a TODO to load it from the AWS default config if AWS API calls are added in here

mselim00 commented on test/manifests/assets/cloudwatch-agent.yaml:
```diff
@@ -0,0 +1,103 @@
+          - {action: replace, target_label: TeamIdentifier, replacement: "{{.TEAM_IDENTIFIER}}"}
+          - {action: replace, target_label: TestType, replacement: "{{.TEST_NAME}}"}
+---
+apiVersion: apps/v1
```
curious how some of these values were set b/c some of them differ from the guide in 
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-metrics.html

particularly things like the requests

annapoornanarayan replied:
Correct me if I am wrong about which values but are you referring to but for [this]( https://github.com/aws/aws-k8s-tester/pull/649/files#diff-1d92a22eaaf79683f108eebbc08d6196a5aaa362c263a7cb29a704888a8be27cR83:~:text=limits%3A,memory%3A%20200Mi ) I referred to our EKSNodeMonitoringAgent cw-agent config [here](https://code.amazon.com/packages/EKSNodeMonitoringAgent/blobs/mainline/--/e2e/monitoring/cloudwatch-agent.yaml ):

annapoornanarayan commented:
Thanks for the comments! Made the following changes:

- TO-DO comment for using kubeconfig for region
- Separate function for rendering CW manifest -> will be useful for nvidia and nvidia-inference tests as they follow a very similar setup
- Added a check for nodes existing before extracting labels
- Better log statement descriptions

annapoornanarayan commented:
- Consolidated all CW related dimension flags into `--metricDimensions` flag which takes in key value pairs. This flag remains in the test phase 
Example:
 ``` 
./nvidia-training.test \
    --test.timeout=60m \
    --test.v \
    --test.run=TestBertTraining \
    --efaEnabled=true \
    --bertTrainingImage=031349313576.dkr.ecr.us-west-2.amazonaws.com/nvidia-test:latest \
    --nodeType=p4d.24xlarge \
    --metricDimensions=TeamIdentifier=node.runtime,kubernetesVersion=1.32,amiVariant=al2023,instanceType=p4d.24xlarge,testType=nvidia.training 
```
- Unlike before there is no validation/hardcoding for these values. They are entirely upto the user and will appear as the Dimensions under `ContainerInsights/Prometheus`
- Moved the Region scraping function, CloudWatch rendering function and `--metricDimension` parsing function to the `raw.go` file so it can be easily accessed by other tests.
- Used string for parsing metric dimension because using a map involves implementing the `flag.Value` interface, which felt excessive for only one flag
- Used HTML to prevent parsing HTML chars as client.go uses `html/template` in RenderManifests. (Using raw strings leads to such strings in output`: &#34;TeamIdentifier&#34;`)
- Used this `fmt.Sprintf("          - {action: replace, target_label: %s, replacement: '%s'}", key, value)) ` to populate. Due to multiline requirement not sure how else to deal with the indentation necessity but open to suggestions.
- Modified Cloudwatch Agent YAML to allow for dynamic dimensions to be set during runtime

annapoornanarayan commented:
Summary of changes: 

- Added common init() for metricDimension flag in `common/config.go`. Used plag package implementation to directly use a map in place of string parsing:
Used `gpflag` with bridging to `flag.CommandLine` for e2e framework compatibility. The e2e framework's `envconf.NewFromFlags()`requires standard Go flags, while `gpflag/pflag` create separate flag sets. Bridging resolves this incompatibility.
- From above, defined MetricDimension as `map[string]string` type
- Enhanced RenderCloudWatchManifest with proper dimension headers (quotes/commas) and simplified templating for relabeling section + removed region requirements
- Refactored global variables in vars.go to struct format (required for `gpflag` initialization)
   Updated variable naming in `main_test.go` and `bert_training_test.go`

mselim00 commented on test/cases/common/config.go:
```diff
@@ -0,0 +1,24 @@
+//go:build e2e
```
test utils like this fit more cleanly into `internal/e2e`, `test/cases` is intended for test executables

mselim00 commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -35,22 +36,30 @@ func TestMain(m *testing.M) {
 
-	manifests := [][]byte{
+	// Render CloudWatch Agent manifest with dynamic dimensions
+	renderedCloudWatchAgentManifest, err := manifests.RenderCloudWatchAgentManifest(common.MetricDimensions)
+	if err != nil {
+		log.Printf("Warning: failed to render CloudWatch Agent manifest: %v", err)
+	}
```
IIUC we don't want to make these requirements, so I just want to sanity check that neither this nor the cloud watch agent daemonset deployment calls are blocking

so 2 qq's:
1. would the CW agent daemonset still go ready even when the infra from the deployer isn't deployed? 
2. can the render function only error if metric dimensions is set and there's some sort of bad value in there?

annapoornanarayan replied:
1. If infra is not deployed, the CW Agent daemonset will fail as it won't find the service account cwagent. Just tested this and it will block the test. I will add a case to make sure it just gives a warning instead
2. Invalid metric dimension values will not cause the render function to fail. The function only performs basic string operations (adding quotes and commas) to format dimension keys. Invalid dimensions may result in malformed output (ex:  [[]])but this still produces valid YAML that creates a functional ConfigMap. Pods will run normally but may display unexepcted or missing dimensions in the CloudWatch dashboard but this does not cause pod failures or affect test execution.

annapoornanarayan commented on test/manifests/assets/dcgm-exporter.yaml:
```diff
@@ -57,6 +59,7 @@ metadata:
     app.kubernetes.io/name: "dcgm-exporter"
     app.kubernetes.io/version: "4.1.3"
 spec:
+  clusterIP: "None"
```
yes exactly! Setting clusterIP:None lets us access the pod IPs directly

annapoornanarayan commented on test/manifests/raw.go:
```diff
@@ -20,4 +33,39 @@ var (
+	CloudWatchAgentManifest []byte
 )
+
+// RenderCloudWatchAgentManifest renders the CloudWatch Agent manifest with dynamic dimensions
+func RenderCloudWatchAgentManifest(metricDimensions map[string]string) ([]byte, error) {
+	var keys []string
+	for key := range metricDimensions {
+		keys = append(keys, `"`+key+`"`)
+	}
+	dimensionsStr := strings.Join(keys, ", ")
+	return fwext.RenderManifests(CloudWatchAgentManifest, map[string]interface{}{
+		"MetricDimensions": metricDimensions,
+		"DimensionKeys":    template.HTML(dimensionsStr),
+	})
+}
+
+// DeployDaemonSet returns a function to deploy and wait for a DaemonSet to be ready
+func DeployDaemonSet(name, namespace string) env.Func {
+	return func(ctx context.Context, config *envconf.Config) (context.Context, error) {
+		log.Printf("Waiting for %s daemonset to be ready.", name)
+		daemonset := appsv1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace},
+		}
+		err := wait.For(
+			fwext.NewConditionExtension(config.Client().Resources()).DaemonSetReady(&daemonset),
+			wait.WithTimeout(5*time.Minute),
+		)
+		if err != nil {
+			return ctx, fmt.Errorf("%s daemonset is not ready: %w", name, err)
+		}
+		log.Printf("%s daemonset is ready.", name)
+		return ctx, nil
+	}
```
Will add it!

annapoornanarayan commented on test/cases/common/config.go:
```diff
@@ -0,0 +1,24 @@
+	}
+	flags.StringToStringVar(&MetricDimensions, "metricDimensions", nil, "Metric dimensions as key=value pairs")
+
+	flags.VisitAll(func(pf *pflag.Flag) {
```
I made the following change in the latest commit, let me know if it follows what you were expecting: I defined the MetricDimensions field as `map[string]string` in the test struct to store the parsed values, then manually registered the flag in the parsing function using `pflag.StringToStringVar()` (because gpflag cannot auto-generate flags for map[string]string types from struct tags). So the flag is now test specific but gets parsed and populated in flags without being a global variable.

annapoornanarayan commented:
Summary

- Renamed `config.go `to` flags.go` and moved to` internal/e2e`
- Moved manifest helpers to `common/helper.go `
- Added case in deploydaemonset to not throw error for optional dameonsets (i.e. dcgm and cloudwatch manifests)
- Added `WithContext` to error logs and changed Config var to zero value equivalent
- Moved MetricDimensions from a  global variable into the Config struct to group test flags, then updated `InitFlags()` to accept a reference to the struct field to populate when parsing

wwvela commented on test/cases/nvidia-training/main_test.go:
```diff
@@ -26,6 +27,7 @@ import (
 )
 
 func TestMain(m *testing.M) {
+	fwext.InitFlags(&testConfig, &testConfig.MetricDimensions)
```
This looks weird. I don't think this is what Nick recommended. You can create a new struct in flag.go like 
```
type MetricOps struct {
        MetricDimensions  map[string]string `flag:"metricDimensions" desc:"CloudWatch metric dimensions as comma-separated key=value pairs"`
}
```
And embed it into Config struct
```
type Config struct {
	MetricOps
	BertTrainingImage string            `flag:"bertTrainingImage" desc:"Docker image used for BERT training workload"`
	EfaEnabled        bool              `flag:"efaEnabled" desc:"Enable Elastic Fabric Adapter (EFA)"`
	NodeType          string            `flag:"nodeType" desc:"Instance type for cluster nodes"`
}
```
For the InitFlags you just need to pass testConfig. And inside of the InitFlags method, it can be something like below. So it can compatible with both config with and without MetricOps embeded. 
```
func InitFlags(config interface{}) error {
    flags, err := gpflag.Parse(config)
    if err != nil {
        return fmt.Errorf("failed to parse flags: %w", err)
    }

    # check if config has MetricOps embeded and init if needed
    InitMetricOps()

    flags.VisitAll(func(pf *pflag.Flag) {
        flag.CommandLine.Var(pf.Value, pf.Name, pf.Usage)
    })

    return nil
}
```
You can simply check if a config has metricOps struct embed [using reflect something like this](https://go.dev/play/p/DAUU0uLz4vY).

annapoornanarayan commented on test/common/helper.go:
Do we like resources.go?

annapoornanarayan commented on internal/e2e/flags.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+// For CloudWatch metric dimension flag
+type MetricOps struct {
```
I moved it as per [this](https://github.com/aws/aws-k8s-tester/pull/649#discussion_r2240834077) comment.
